### PR TITLE
Complete shopping cart task

### PR DESCRIPTION
**Pull request by Stelmaschyk**

### CODE REVIEW

File: src/main/java/mate/academy/bookstore/controller/ShoppingCartController.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+@RequiredArgsConstructor
+@RestController
+@RequestMapping("/cart")
+public class ShoppingCartController {
+    private final ShoppingCartService shoppingCartService;
+
+    @PostMapping
+    @ResponseStatus(HttpStatus.OK)
+    @Operation(summary = "Add books to cart", description = "Add books to the shopping cart")
+    public CartItemResponseDto addBookToCart(@RequestBody @Valid CartItemRequestDto request,
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/controller/ShoppingCartController.java]:
```suggestion
    public ShoppingCartResponseDto addBookToCart(@RequestBody @Valid CartItemRequestDto request,
```

File: src/main/java/mate/academy/bookstore/controller/ShoppingCartController.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    public ShoppingCartResponseDto getCart(Authentication authentication) {
+        User user = (User) authentication.getPrincipal();
+        return shoppingCartService.getShoppingCart(user.getEmail());
+    }
+
+    @PutMapping("/items/{id}")
+    @ResponseStatus(HttpStatus.OK)
+    @Operation(summary = "Update quantity in a cart", description = "Update books quantity in "
+            + "the shopping cart")
+    public CartItemResponseDto updateCartItem(@RequestBody @Valid CartItemUpdateDto request,
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/controller/ShoppingCartController.java]:
```suggestion
    public ShoppingCartResponseDto updateCartItem(@RequestBody @Valid CartItemUpdateDto request,
```

File: src/main/java/mate/academy/bookstore/mapper/ShoppingCartMapper.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+import mate.academy.bookstore.model.ShoppingCart;
+import org.mapstruct.AfterMapping;
+import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
+import org.mapstruct.MappingTarget;
+import org.mapstruct.ReportingPolicy;
+
+@Mapper(config = MapperConfig.class, unmappedTargetPolicy = ReportingPolicy.IGNORE)
+public interface ShoppingCartMapper {
+
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/mapper/ShoppingCartMapper.java]: remove empty line

File: src/main/java/mate/academy/bookstore/model/CartItem.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.Where;
+
+@Entity
+@Table(name = "cart_items")
+@Getter
+@Setter
+@SQLDelete(sql = "UPDATE cart_items SET is_deleted = true WHERE id=?")
+@Where(clause = "is_deleted=false")
+@RequiredArgsConstructor
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/CartItem.java]:
```suggestion
```

File: src/main/java/mate/academy/bookstore/model/CartItem.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+public class CartItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @JoinColumn(name = "shopping_cart_id",nullable = false)
+    @ManyToOne
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/CartItem.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

File: src/main/java/mate/academy/bookstore/model/CartItem.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+@Getter
+@Setter
+@SQLDelete(sql = "UPDATE cart_items SET is_deleted = true WHERE id=?")
+@Where(clause = "is_deleted=false")
+@RequiredArgsConstructor
+public class CartItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @JoinColumn(name = "shopping_cart_id",nullable = false)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/CartItem.java]:
```suggestion
    @JoinColumn(name = "shopping_cart_id", nullable = false)
```

File: src/main/java/mate/academy/bookstore/model/CartItem.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    @ManyToOne
+    private ShoppingCart shoppingCart;
+    @JoinColumn(name = "book_id",nullable = false)
+    @OneToOne
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/CartItem.java]:
```suggestion
    @OneToOne(fetch = FetchType.LAZY, optional = false)
```

File: src/main/java/mate/academy/bookstore/model/CartItem.java [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+@Where(clause = "is_deleted=false")
+@RequiredArgsConstructor
+public class CartItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @JoinColumn(name = "shopping_cart_id",nullable = false)
+    @ManyToOne
+    private ShoppingCart shoppingCart;
+    @JoinColumn(name = "book_id",nullable = false)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/CartItem.java]:
```suggestion
    @JoinColumn(name = "book_id", nullable = false)
```

File: src/main/java/mate/academy/bookstore/model/ShoppingCart.java [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.SQLRestriction;
+
+@Entity
+@Getter
+@Setter
+@SQLDelete(sql = "UPDATE shopping_carts SET is_deleted = true WHERE id=?")
+@SQLRestriction("is_deleted=false")
+@Table(name = "shopping_carts")
+@RequiredArgsConstructor
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/ShoppingCart.java]:
```suggestion
```

File: src/main/java/mate/academy/bookstore/model/ShoppingCart.java [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+public class ShoppingCart {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @JoinColumn(name = "user_id", nullable = false)
+    @OneToOne
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/ShoppingCart.java:32]: You can use `@MapsId` here

File: src/main/java/mate/academy/bookstore/model/ShoppingCart.java [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    private User user;
+    @Column(nullable = false)
+    private boolean isDeleted = false;
+    @ToString.Exclude
+    @EqualsAndHashCode.Exclude
+    @OneToMany(mappedBy = "shoppingCart")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/ShoppingCart.java]:
```suggestion
    @OneToMany(mappedBy = "shoppingCart", cascade = CascadeType.ALL, orphanRemoval = true)
```

File: src/main/java/mate/academy/bookstore/model/ShoppingCart.java [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+    private Set<CartItem> cartItems;
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/ShoppingCart.java]:
```suggestion
    private Set<CartItem> cartItems = new HashSet<>();
```

File: src/main/java/mate/academy/bookstore/service/ShoppingCartServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+        shoppingCartRepository.save(shoppingCart);
+    }
+
+    @Override
+    public ShoppingCartResponseDto getShoppingCart(String email) {
+        return shoppingCartMapper.toDto(shoppingCartRepository.findByUserEmail(email));
+    }
+
+    @Override
+    public CartItemResponseDto addCartItem(User user, CartItemRequestDto request) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/service/ShoppingCartServiceImpl.java]:
```suggestion
    public ShoppingCartResponseDto addCartItem(User user, CartItemRequestDto request) {
```

File: src/main/java/mate/academy/bookstore/service/ShoppingCartServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+            }
+        }
+        CartItem cartItem = cartItemMapper.toModel(request, book);
+        cartItem.setBook(book);
+        cartItem.setShoppingCart(shoppingCart);
+        return cartItemMapper.toDto(cartItemRepository.save(cartItem));
+    }
+
+    @Override
+    public CartItemResponseDto updateBooksQuantity(User user, Long cartItemId,
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/service/ShoppingCartServiceImpl.java]:
```suggestion
    public ShoppingCartResponseDto updateBooksQuantity(User user, Long cartItemId,
```

File: src/main/java/mate/academy/bookstore/service/ShoppingCartServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+        CartItem cartItem = cartItemRepository.findById(cartItemId).orElseThrow(
+                    () -> new EntityNotFoundException("not found cart item by id: " + cartItemId));
+        cartItem.setQuantity(request.getQuantity());
+        return cartItemMapper.toDto(cartItemRepository.save(cartItem));
+    }
+
+    @Override
+    public void deleteById(Long id) {
+        cartItemRepository.deleteById(id);
+    }
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/service/ShoppingCartServiceImpl.java:67]: Think where `@Transactional` should be

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/mate/academy/bookstore/model/CartItem.java (hunk 5 above)
Comment (Stelmaschyk) [reply src/main/java/mate/academy/bookstore/model/CartItem.java]: Thanks for new knowledge! Its about "optional = false"

File: src/main/java/mate/academy/bookstore/model/ShoppingCart.java (hunk 11 above)
Comment (Stelmaschyk) [reply src/main/java/mate/academy/bookstore/model/ShoppingCart.java]: Thanks for new knowledge! Its about "orphanRemoval = true"

File: src/main/resources/db/changelog/changes/10-create-shopping_carts-table.yaml [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+                  defaultValueBoolean: false
+                  constraints:
+                    nullable: false
+
+        - addForeignKeyConstraint:
+            baseTableName: shopping_carts
+            baseColumnNames: user_id
+            constraintName: fk_shopping_cart_user
+            referencedTableName: users
+            referencedColumnNames: id
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/changes/10-create-shopping_carts-table.yaml]: add one empty line at the end of the file ;) 
check other files

**Review by liliia-ponomarenko — APPROVED** [review]
good job! 👍

**Pull request merged**